virttest/qemu_vm.py

```diff
@@ -1106,8 +1106,41 @@ def add_qxl(qxl_nr, base_addr=29):
                 qxl_str += " -device qxl,id=video%d,addr=0x%x" % (index, addr)
             return qxl_str
 
-        def add_vga(vga):
-            return " -vga %s" % vga
+        def add_vga(devices, vga):
+            """Add primary vga device."""
+            default_machine_re = re.compile(".*")
+            isapc_machine_re = re.compile("isapc")
+            pci_bus = {'aobject': 'pci.0'}
+            isa_bus = None  # so far there is no ISA bus
+            vga_dev_map = {
+                # name: [(machine regex, driver, parent bus), ...]
+                # the default variant must be the last one
+                "std": [(isapc_machine_re, "isa-vga", isa_bus),
+                        (default_machine_re, "VGA", pci_bus)],
+                "cirrus": [(isapc_machine_re, "isa-cirrus-vga", isa_bus),
+                           (default_machine_re, "cirrus-vga", pci_bus)],
+                "vmware": [(default_machine_re, "vmware-svga", pci_bus)],
+                "qxl": [(default_machine_re, "qxl-vga", pci_bus)],
+                "virtio": [(default_machine_re, "virtio-vga", pci_bus)],
+            }
```

Comment:
This adds quite a bit of overhead. If you insist on mapping, how about doing it conditional only on non-default machine-types (in general they are not used that often and the mapping will be way simpler):

```python
bus = {"aobject": "pci.0"}
vga_dev_map = {"std": "VGA", "cirrus": "cirrus-vga", "vmware": "vmware-svga", ...}

vga_dev = vga_dev_map.get(vga, vga_dev)
if (not vga_dev or
        params.get("vga_use_legacy_expression")):
    fallback = True
else:
    if '-mmio:' in machine_type:
        bus = None
        if vga_dev.endswith('-pci'):
            vga_dev = "%s%s" % (vga_dev[:-4], -device)
        else:
            fallback = True
    elif machine_type == "isapc":
        bus = None
        if vga == "std":
            ...
        elif vga == "cirrus":
            ...
    else:
        fallback = False
    if (not fallback and not devices.has_device(vga_dev)):
        fallback = True

if fallback:
    return StrDevice()
else:
    return Device()
```

This would:

* not allow negative testing, but that could be done with "vga=None" and "extra_params"
* allow specifying isa/pci/mmio VGAs
* allow specifying custom vgas (`vga=foo` in case "foo" exists would use it)
* won't do any compiling/matching in default scenario (pci)
* simple matching of pci-devices to isa/mmio in case of isa/mmio machine (with priority according to frequency of use, btw is anyone using isapc?)

Reply:
Honestly, I don't like the above implementation, it took me a while to correctly understand how it works. Let me translate to another version to help people understand it, the code would be:
```python
def add_vga_machine_default(vga):
    ...

def add_vga_machine_mmio(vga):
    ...

def add_vga_machine_isapc(vga):
    ...

if '-mmio:' in machine_type:
    add_vga_machine_mmio(vga)
elif machine_type == "isapc":
    add_vga_machine_isapc(vga)
else:
    add_vga_machine_default(vga)
```
so, if the requirement is to make the implementation in an object-oriented way (or something like what I showed here), I would be happy to make the change, otherwise, I'd prefer my current one as it has good readability.

Reply:
well none of these has anything remotely related to object-oriented style. Sure, I optimized it which might decreased the readability, but without the optimization it's:

```python
# Get default avocado->qemu mapping
vga_dev = vga_dev_map.get(vga, vga_dev)

# Check per-machine-type mapping
if mmio:
    ...
elif isapc:
    ...

if fallback:
    return
else:
    return
```

If you want to go OO style, then please split QContainer implementation per-machine-type so it can use different method per each machine type (if necessary). Anyway the result will be the same:

```
class QPCIContainer:
    def add_vga(self, vga):
        vga = vga_dev_map.get(vga, vga_dev)
        ...

class QMMIOContainer:
    def add_vga(self, vga):
        vga = super(QMMIOContainer, self).add_vga(vga)
        vga = pci_to_mmio_map.get(vga)
```

(what I mean is the interface will be `add_vga(avocado-vga-type)` and the process would be `map_avocado_vga_to_qemu_vga` and on some platforms followed by `map_pci_qemu_vga_to_my_type_of_vga`).